validation/validation_runner.py

```diff
@@ -0,0 +1,359 @@
+
+    Returns
+    -------
+    Path to the truth VCF and confident region BED file
+    """
+
+    a_query_model = AnalysisQueryModel(
+        projects=[get_config()['workflow']['dataset']],
+        sample_ids=[cpg_id],
+        type=AnalysisType('custom'),
+        meta={'type': 'validation'},
+        active=True,
+    )
+    analyses = AnalysisApi().query_analyses(analysis_query_model=a_query_model)
+    if len(analyses) > 1:
+        raise Exception(
+            f'Multiple [custom] analysis objects were found for '
+            f'{cpg_id}, please set old analyses to active=False'
+        )
+
+    if len(analyses) == 0:
+        raise Exception(f'{cpg_id} has no Analysis entries')
+
+    analysis_object = analyses[0]
+    truth_vcf = analysis_object.get('output')
+    truth_bed = analysis_object['meta'].get('confident_region')
+    assert (
+        truth_bed and truth_vcf
+    ), f'Missing one or both of the truth files: BED: {truth_bed}, VCF: {truth_vcf}'
+
+    return truth_vcf, truth_bed
+
+
+def post_results_job(
+    batch: hb.batch.Batch,
+    dependency: hb.batch.job.Job,
+    sample_id: str,
+    ss_vcf: str,
+    truth_vcf: str,
+    truth_bed: str,
+    joint_mt: str,
+):
+    """
+    post the results to THE METAMIST using companion script
+
+    Parameters
+    ----------
+    batch : batch to run the job in
+    dependency : the analysis job we're dependent on
+    sample_id : CPG ID
+    ss_vcf : single sample VCF file used/created
+    truth_vcf : source of truth variants
+    truth_bed : source of confident regions
+    joint_mt : joint-call MatrixTable
+    """
+
+    post_job = batch.new_job(name=f'Update metamist for {sample_id}')
+    post_job.depends_on(dependency)
+    post_job.image(get_config()['workflow']['driver_image'])
+    copy_common_env(post_job)
+    authenticate_cloud_credentials_in_job(post_job)
+    job_cmd = (
+        f'PYTHONPATH=$(pwd) python3 {RESULTS_SCRIPT} '
+        f'--id {sample_id} '
+        f'--ss {ss_vcf} '
+        f'-t {truth_vcf} '
+        f'-b {truth_bed} '
+        f'--mt {joint_mt} '
+    )
+    post_job.command(job_cmd)
+
+
+def main(input_file: str, header: str | None):
+    """
+    Parameters
+    ----------
+    input_file : path to the MT representing this joint-call
+    header : any additional header lines to add when writing VCF
+    """
+
+    validation_lookup = get_validation_samples()
+
+    service_backend = hb.ServiceBackend(
+        billing_project=get_config()['hail']['billing_project'],
+        remote_tmpdir=remote_tmpdir(),
+    )
+    batch = hb.Batch(
+        name='run validation bits and pieces',
+        backend=service_backend,
+        cancel_after_n_failures=1,
+        default_memory='highmem',
+    )
+
+    input_path = Path(input_file)
+    if input_path.suffix != '.mt':
+        logger.error('Expected a MT as input file')
+        raise Exception('Expected a MT as input file')
+
+    # requires conversion to vcf
+    mt_to_vcf(
```

Comment:
Should it be submitted as a batch job?

Reply:
Yeah, I think this could do with a re-write. It was done this way so that this stage (creating all single sample VCFs) would be completed before the next stage (for each single sample VCF we created, find the truth data and run validation)

This could easily be rewritten as:
- find the samples in the validation project from metamist
- open the MT to check which samples are in the joint call
- for each intersecting sample:
    1. check there is a truth VCF/BED in Metamist
    2. create single sample VCF (job)
    3. validate against the truth (job)
    4. log results back into metamist (job)

Reply:
If this is ever integrated with the pipeline we could take all relevant samples as an argument, and save a few more steps

Reply:
I've updated it to follow this approach ^^ 

Now single-sample VCFs are generated using a batch job per sample, and downstream comparison and posting results are dependent on the job specific to that sample (or no VCF dependency if the file didn't need to be recreated)

Reply:
> If this is ever integrated with the pipeline we could take all relevant samples as an argument, and save a few more steps

I'm happy to look into that! We can just use the `get_cohort()` function from cpg-utils :)